### PR TITLE
Performance tests: Refactor use of ENV to simplify test runner logic

### PR DESCRIPTION
**Pull request by dmsnell**
## What?

Refactors and de-abstracts the performance test runner, relies on ENV values set by GitHub Actions to configure test branches for PR test runs.

## Why?

As part of an ongoing effort to optimize the performance tests this change de-abstracts some of the interface to the runner script and refactors to minimize future changes which impact how we checkout/clone the referenced `git` repository and branches.

By rearranging the code here we can more easily take advantage of situational advantages when checking out code, such as reusing the existing local `git` checkout provided by earlier steps in the GitHub Actions job.

There are several new semantic terms introduced to replace implicit logical conditions in hope of making the flow through the test suite clearer; this is important because the test suite currently serves at least three separate purposes: running on PRs to predict impact of new changes; running after PR merge into `trunk` to measure historical performance trends; and running against major releases to track impact on performance.

## How?

 - Removes CLI args that can be inferred from ENV values set by Github Actions
 - Consumes a `git` client directly instead of using the `git` abstraction in the test utils, making it clearer what exactly is happening and making it easier to modify that behavior.

---

There should be no functional or performance changes in this PR. This is a code refactoring in anticipation of further changes which should bring positive impact on test runtime.

### CODE REVIEW

File: bin/plugin/commands/performance.js [hunk 1]
```diff
@@ -193,7 +175,11 @@ async function runTestSuite( testSuite, performanceTestDirectory ) {
  * @param {WPPerformanceCommandOptions} options  Command options.
  */
 async function runPerformanceTests( branches, options ) {
+	const branchesWereSpecified = branches?.length > 0;
 	const runningInCI = !! process.env.CI || !! options.ci;
+	const inferTestBranches = runningInCI && ! branchesWereSpecified;
+	const mergeRef = process.env.GITHUB_SHA ?? '';
+	const baseRef = process.env.GITHUB_BASE_REF ?? '';
```
Comment (youknowriad) [inline bin/plugin/commands/performance.js]: Personally I don't like that this PR ties the performance command with GitHub actions. I think a command to just pass a random number of branches/commits is way more understandable from an API perspective. 

Is it possible to leave this logic to figure out the right branch out of the command itself? potentially a new command?

Comment (dmsnell) [reply bin/plugin/commands/performance.js]: Thanks @youknowriad. I removed those in 316983f53b9

They were an artifact of my explorations and I overlooked the fact that I don't need them anymore because of values available in the `ENV` that I was previously getting through `git` itself.

File: bin/plugin/commands/performance.js [hunk 2]
```diff
@@ -214,24 +196,47 @@ async function runPerformanceTests( branches, options ) {
...
+	const git = SimpleGit( baseDirectory );
+	await git
+		.raw( 'init' )
+		.raw( 'remote', 'add', 'origin', config.gitRepositoryURL );
+
+	for ( const branch of branches ) {
+		await git.raw( 'fetch', '--depth=1', 'origin', branch );
+	}
+
+	await git.raw( 'checkout', branches[ 0 ] );
```
Comment (youknowriad) [inline bin/plugin/commands/performance.js:221]: How is the logic here different than the `git clone` call that was removed?

File: bin/plugin/commands/performance.js [hunk 3]
```diff
@@ -214,24 +196,47 @@ async function runPerformanceTests( branches, options ) {
...
-			performanceTestDirectory,
-			options.testsBranch
-		);
+		const branchName = formats.success( options.testsBranch );
+		log( `    >> Fetching the test-runner branch: ${ branchName }` );
+
+		// @ts-ignore
+		await SimpleGit( performanceTestDirectory )
+			.raw( 'fetch', '--depth=1', 'origin', options.testsBranch )
+			.raw( 'checkout', options.testsBranch );
```
Comment (youknowriad) [inline bin/plugin/commands/performance.js:237]: I guess I have the same question here :). It seems you're moving away from the "git" abstraction we had but I'm not understanding the "why"

File: bin/plugin/commands/performance.js (hunk 2 above)
Comment (dmsnell) [reply bin/plugin/commands/performance.js:221]: one why reason I pulled out the custom `git` abstraction was because I found it hard to follow what was actually going on by jumping around the various functions and modules,  and then I'm not comfortable making changes to the abstraction given my limited understanding of what else relies on it.

the big change here is that all fetches are shallow and that we're not passing `--no-single-branch`. I'm guessing, because there's no record in the codebase for why  `--no-single-branch` was added, that someone added it when one of the referenced branches didn't appear in the shallow clone. however, that option is going to pull in around 1700 commits right now based on the repository today and that amounts to a jump from roughly 20MB downloaded in the clone for what we need to 190MB for all the other things we don't. that also accounts for several minutes of runtime for the job.

later in #45327 (where I'm collecting the set of successful experiments) I'd like to first fetch from the copy of the repo already cloned in previous step and avoid making network calls during the job run. this work alone accounts for around six minutes of runtime and cuts in half the variability of runtime duration in half.

because the `git` abstraction creates directories on clone, because it creates its own `git` object on every call, and because it doesn't return that `git` object we can't easily make minor adjustments in how we are using `git`, which has made these kinds of changes hard to setup or code without pulling in `simple-git` anyway in addition to the abstraction.

Comment (dmsnell) [reply bin/plugin/commands/performance.js:221]: one why reason I pulled out the custom `git` abstraction was because I found it hard to follow what was actually going on by jumping around the various functions and modules,  and then I'm not comfortable making changes to the abstraction given my limited understanding of what else relies on it.

the big change here is that all fetches are shallow and that we're not passing `--no-single-branch`. I'm guessing, because there's no record in the codebase for why  `--no-single-branch` was added, that someone added it when one of the referenced branches didn't appear in the shallow clone. however, that option is going to pull in around 1700 commits right now based on the repository today and that amounts to a jump from roughly 20MB downloaded in the clone for what we need to 190MB for all the other things we don't. that also accounts for several minutes of runtime for the job.

later in #45327 (where I'm collecting the set of successful experiments) I'd like to first fetch from the copy of the repo already cloned in previous step and avoid making network calls during the job run. this work alone accounts for around six minutes of runtime and cuts in half the variability of runtime duration.

because the `git` abstraction creates directories on clone, because it creates its own `git` object on every call, and because it doesn't return that `git` object we can't easily make minor adjustments in how we are using `git`, which has made these kinds of changes hard to setup or code without pulling in `simple-git` anyway in addition to the abstraction.

Comment (youknowriad) [reply bin/plugin/commands/performance.js:221]: Should we remove it entirely. I guess it's still used by one other command?

Comment (dmsnell) [reply bin/plugin/commands/performance.js:221]: I would be happy to remove it entirely, as I think that its main contribution is a layer of abstraction that adds indirection while failing to aid in comprehension or efficiency. Generally I'm hesitant to remove things for which I don't fully understand their use or motivation, but I could extend this PR to remove it from the other place also.

Comment (dmsnell) [reply bin/plugin/commands/performance.js:221]: removed in 0445494162c

it's a bigger change than I prefer, and I plan to add some explanatory comments on things that stood out to me while refactoring (such as why we're using `git pull` to "rest" a branch)

I will request a more careful review than normal to make sure I didn't change anything in the port; already I caught myself accidentally using `trunk` instead of `npmBranchName` in one place.

**Comment by dmsnell** [comment]
Can someone familiar with project releases give this a look-over? I'm not entirely sure how to verify that this doesn't accidentally break package and plugin releases.

File: bin/plugin/commands/packages.js [hunk 4]
```diff
@@ -105,13 +139,19 @@ async function runNpmReleaseBranchSyncStep( pluginReleaseBranch, config ) {
...
-			gitWorkingDirectoryPath,
-			pluginReleaseBranch
-		);
+		const repo = git( gitWorkingDirectoryPath );
+
+		/*
+		 * Replace content from remote branch.
+		 *
+		 * @TODO: What is our goal here? Could `git reset --hard origin/${pluginReleaseBranch}` work?
+		 *        Why are we manually removing and then adding files back in?
```
Comment (ockham) [inline bin/plugin/commands/packages.js:148]: > What is our goal here? Could `git reset --hard origin/${pluginReleaseBranch}` work?

I was puzzled by the same question when I last touched this file. Let's give your suggestion a try :+1:

> Why are we manually removing and then adding files back in?

I think it's mostly to cover any files that were added and/or removed with respect to the other branch?

File: bin/plugin/commands/packages.js [hunk 5]
```diff
@@ -439,18 +475,22 @@ async function backportCommitsToBranch(
...
-		branchName
-	);
+	const repo = git( gitWorkingDirectoryPath );
+
+	/*
+	 * Reset any local changes and replace them with the origin branch's copy.
+	 *
+	 * @TODO: Why do we use fetch/checkout/pull here instead of checkout/reset--hard?
+	 *        Is the fetch necessary at all? Do we expect the remote branch to be
+	 *        different now than when we started running this script?
```
Comment (ockham) [inline bin/plugin/commands/packages.js]: We've had some issues that I believe were due to a scenario like that, see https://github.com/WordPress/gutenberg/pull/30615. I'd lean towards keeping the fetch; let's maybe update the comment.

File: bin/plugin/commands/packages.js (hunk 4 above)
Comment (dmsnell) [reply bin/plugin/commands/packages.js:148]: @ockham do you think it would be fair to do this in a follow-up PR or make the change here? I'm nervous about smushing the logic change with the refactor in case it turns out that it was necessary to be the way it is.

Comment (ockham) [reply bin/plugin/commands/packages.js:148]: Makes sense. Follow-up is fine :+1:

File: bin/plugin/commands/packages.js (hunk 5 above)
Comment (dmsnell) [reply bin/plugin/commands/packages.js]: Ah thank you. That makes sense, especially here where we're back-porting the changes and need to push something fast-forwardable. I'll add a note.

File: bin/plugin/commands/performance.js [hunk 6]
```diff
@@ -214,24 +196,47 @@ async function runPerformanceTests( branches, options ) {
...
+	}
+
+	const baseDirectory = getRandomTemporaryPath();
+	fs.mkdirSync( baseDirectory, { recursive: true } );
+
+	// @ts-ignore
+	const git = SimpleGit( baseDirectory );
+	await git
+		.raw( 'init' )
+		.raw( 'remote', 'add', 'origin', config.gitRepositoryURL );
```
Comment (ockham) [inline bin/plugin/commands/performance.js:215]: Could use `await` [`git.init()`](https://github.com/steveukx/git-js#git-init)[`.addRemote( 'origin', config.gitRepositoryURL )`](https://github.com/steveukx/git-js#git-remote), but I guess you've covered that [below](https://github.com/WordPress/gutenberg/pull/45255/files#r1008373526).

File: bin/plugin/commands/performance.js [hunk 7]
```diff
@@ -214,24 +196,47 @@ async function runPerformanceTests( branches, options ) {
...
+
+	for ( const branch of branches ) {
+		await git.raw( 'fetch', '--depth=1', 'origin', branch );
```
Comment (ockham) [inline bin/plugin/commands/performance.js:218]: `await git.fetch( origin, branch, [ '--depth=100' ] );`

File: bin/plugin/commands/performance.js (hunk 6 above)
Comment (dmsnell) [reply bin/plugin/commands/performance.js:215]: thanks. I tried changing it but there's a thing I don't like about this pattern which is that `SimpleGit` is making us guess which functions it has chosen to abstract and which not to, and then we need to guess what that abstraction does.

case in point, we have `SimpleGit.addRemote()` but not `SimpleGit.showRemote()`. So while `.init()` makes sense, I feel like the raw version of `remote add` does far less to obscure what's going on, even if in the end it does the same thing as raw does.

I'd like to leave this as is for now until and unless we see a compelling reason why using this middleman/new term is more useful than using the more true `git` vocabulary and idioms.

File: bin/plugin/commands/performance.js (hunk 3 above)
Comment (ockham) [reply bin/plugin/commands/performance.js:237]: I had a similar question that Dennis answered [here](https://github.com/WordPress/gutenberg/pull/45255/files#r1018520922).

**Review by ockham — APPROVED** [review]
I've tested publishing GB on my fork, and it seems to work (and run performance tests). LGTM -- let's maybe leave a note in #core-editor to make sure people are aware of this change in case something goes wrong after all next time a release is published.

**Comment by Mamaduka** [comment]
@dmsnell, I started seeing the following errors for Playwright and Performance tests actions:

```
GitError: error: pathspec 'trunk' did not match any file(s) known to git
```

## Screenshot

CleanShot 2022-11-15 at 12 22 37

**Comment by youknowriad** [comment]
@Mamaduka I think that error is the GitHub problem being talked about here  https://wordpress.slack.com/archives/C02QB8GMM/p1668497600444079

**Comment by gziolo** [comment]
I saw the following issue introduced with this refactor:

Screenshot 2023-08-10 at 13 30 47

I'll try to resolve it, as it looks like the config entry is read from a wrong place now.

**Comment by gziolo** [comment]
I resolved the previous issue, but now I see the following:

Screenshot 2023-08-10 at 13 34 58

**Comment by gziolo** [comment]
I fixed everything, I will open PR tomorrow with improvements to the npm publishing task.

**Comment by dmsnell** [comment]
@gziolo please add me if you like on that PR. also if you want any help I'm happy to try and be of service.

**Comment by gziolo** [comment]
The fix is available at https://github.com/WordPress/gutenberg/pull/53565.
